### ISSUE TITLE
Iterative solvers and Composition

### ISSUE THREAD

**Issue by upsj** [ginkgo-project/ginkgo#541]
Right now, the intermediate values used as x in composition-like LinOps (Composition, preconditioner::Ilu) are often uninitialized. We could use `apply_uses_initial_guess` to determine whether we need to initialize this with a sensible value.
For square matrices, we could use b, for non-square matrices, we could zero-initialize. Any other ideas?

**Comment by pratikvn** [comment]
I guess it is not really possible to set a custom user-defined array ? You would have to expose the settings of the intermediate vectors to the user, then ?

**Comment by upsj** [comment]
Yes, and also we can have arbitrarily many of these intermediate vectors, so the only general solution I see is using the input (square matrices) or constant values (non-square matrices)
